\documentclass{article}

\usepackage{multicol}
\usepackage{geometry}
\usepackage{amsmath}

\geometry{
 a4paper,
 total={210mm,297mm},
 left=20mm,
 right=20mm,
 top=20mm,
 bottom=20mm,
 }

\begin{document}

\title{GTEx expression data\\-\\
        Rainet project}
\author{Diogo Ribeiro, Lionel Spinelli, Andreas Zanzoni, Christine Brun}
\maketitle

\section{Introduction} 

We plan to use the GTEx V6 dataset (http://gtexportal.org), Human genome-wide RNA-seq expression data, to add RNA expression information in our RAINET database. 
The purpose is to use the expression to filter out Protein-RNA interactions that unlikely to occur in vivo. We will apply a threshold of expression value to ascertain the potential presence of each protein and RNA in a certain tissue, and excluded interactions where one or both of the interaction partners are missing. 
We will extrapolate the protein presence or absence by the expression of their respective mRNA. It as been shown (REF) that prediction of protein presence is accurate when its RNA is expressed above ..X.. rpkm. \par
The GTEx dataset contains full RNA-seq experiments done on hundreds of individuals, and retrieved from several physical locations / tissues. 
The GTEx project provides downloadable data where the reads have been mapped to human genome and rpkm values were calculated for each GENCODE (v19) transcript (note that this includes all types of RNAs: mRNAs, lncRNAs, snoRNAs, etc).
They provide several files, among them the description of the individual RNA-seq samples, including the tissue and body site:
GTEx\_Data\_V6\_Annotations\_SampleAttributesDS.txt. \par
The file with the rpkm values contains values for each transcript for each sample:
GTEx\_Analysis\_v6\_RNA-seq\_Flux1.6\_transcript\_rpkm.txt. \par
For insertion of expression data into our RAINET database we want to transform the GTEx data to have a single expression value for each RNA-tissue pair, therefore, exclude the sample/individual dimension from the data. 
We will first perform a rough analysis of expression values distributions across samples to reach the best solution for transforming this data.

\section{Samples variability per tissue}

Number of samples per tissue using the whole GTEx\_Data\_V6\_Annotations\_SampleAttributesDS.txt
file with the attribute SMTS (the less-specific tissue terms).

<<samples_per_tissue, fig.keep="all", echo=FALSE, message=F, warning=F>>=
  source( paste( working_dir, "samples_per_tissue.R", sep="/"))
@

We can see that number of samples per tissue is highly variable. Perhaps this should be considered in the further analysis.

\section{Expression variability per tissue}

Distribution of all the expression values per tissue of 50 randomly sampled transcripts (out of >190 thousand), sampled from the whole GTEx rpkm file. Keep in mind that different tissues have different sample sizes, and thus a different amount of values are plotted on each tissue. \par 
The purpose here is not to compare which tissue has higher or lower expression, or higher or lower variability in expression, but to understand the variability of expression between each tissue, coming from both biological variability and sampling variability. \par
The left-side of the following boxplot contains the distribution of expression values for the sample transcripts using all available samples (e.g. thousands in brain, a handful in Fallopian tube), whereas the right-side boxplot displays the data for the same transcript when only considering 6 (random) samples for each tissue.

<<expression_per_tissue, fig.keep="all", echo=FALSE, message=F, warning=F>>=
  source( paste( working_dir, "expression_per_tissue.R", sep="/"))
@

We can see that the number of samples does not drastically change the overal distribution of expression per tissue, only that the number of extreme values (i.e. the black dots) are decresed when reducing number of samples. Thus, as we plan to use some averaging metric between samples of the same tissue for each transcript, we use a different number of samples per tissue, without impairing the credibility or comparibility of the average expression value.

\newpage

\section{Expression distribution per transcript}

Selected 5 tissues with different sample numbers (2 with highest, 1 medium, 2 with lowest sample numbers), randomly selected 12 transcripts and plotted their RPKM value distribution, using all available samples. Note that many transcript have 0 or close to 0 RPKM (*Bug in graphics when all RPKMs are zero) in all addressed tissues. We plan to average the RPKM values of a transcript in a given tissue.

<<density_plot_example_transcripts, fig.keep="all", echo=FALSE, message=F, warning=F>>=
  source( paste( working_dir, "density_plot_example_transcripts.R", sep="/"))
@

The distributions are highly variable between transcripts. 

\section{Average expression of transcripts per tissue}

To merge sample values within the same tissue and transcript, we can use mean or median. Below are the distributions of averaged RPKM values among all transcripts and all tissues.\par
The blue horizontal lines represent either the mean or the median, the red lines represent the noise cutoff of 0.1 RPKM suggested in the GTEx papers.

<<transcript_expression_averages, fig.keep="all", echo=FALSE, message=F, warning=F>>=
  source( paste( working_dir, "transcript_expression_averages.R", sep="/"))
@

\section{Conclusion}

We intend to use this expression data not for an expression analysis, but only as a filter layer to ensure lncRNA-protein co-existence in a cell. After setting a rpkm cutoff that distinguishes expression noise from real expression, we will turn our values into binary presence/absence.\par
Because we want to keep the maximum data possible, we do not need to filter out tissues that have low sample numbers, or to normalise data in tissues with a vast number of samples.\\


Parameters for filtering:\\
- minimum rpkm value\\
- interacting pair having to be present in X\% tissues

Need to decide:\\
- what level: tissue or body part?\\

\end{document}

